Ruby: add docs for customizing library models with data extensions (#15488)
github/codeql: docs/codeql/codeql-language-guides/customizing-library-models-for-ruby.rst

.. _customizing-library-models-for-ruby:

:orphan:
:nosearch:

Customizing Library Models for Ruby
===================================

.. include:: ../reusables/beta-note-customizing-library-models.rst

Ruby analysis can be customized by adding library models in data extension files.

A data extension for Ruby is a YAML file of the form:

.. code-block:: yaml

  extensions:
    - addsTo:
        pack: codeql/ruby-all
        extensible: <name of extensible predicate>
      data:
        - <tuple1>
        - <tuple2>
        - ...

> The CodeQL library for JavaScript exposes the following extensible predicates:

Reviewer: ```suggestion
The CodeQL library for Ruby exposes the following extensible predicates:
```

.. code-block:: ruby

  tty = TTY::Command.new
  tty.run(cmd) # <-- add 'cmd' as a taint sink

For this example, you can use the following data extension:

.. code-block:: yaml

  extensions:
    - addsTo:
        pack: codeql/ruby-all
        extensible: sinkModel
      data:
        - ["TTY::Command", "Method[run].Argument[0]", "command-injection"]

- Since we're adding a new sink, we add a tuple to the **sinkModel** extensible predicate.
- The first column, **"TTY::Command"**, identifies a set of values from which to begin the search for the sink.
  The string **"TTY::Command""** means we start at the places where the codebase constructs instances of the class **TTY::Command**.
- The second column is an access path that is evaluated from left to right, starting at the values that were identified by the first column.

  - **Method[run]** selects calls to the **run** method of the **TTY::Command** class.
  - **Argument[0]** selects the first argument to calls to that member.

- **command-injection** indicates that this is considered a sink for the command injection query.

> Example: Taint sources from `sinatra` block parameters

Reviewer: Minor: I think this will render "sinatra" in italics. You can use double backticks but it might be easier to just use single inverted commas as with the other headers on this page

------------------------------------------------------

> In this example, we'll show how the `x` parameter below could be marked as a remote flow source:

Reviewer: ```suggestion
In this example, we'll show how the 'x' parameter below could be marked as a remote flow source:
```

Another one where I think it'll otherwise render in italics, and would be ideal to use the same format throughout the doc

- **Argument[0]** selects the first argument to the method call.
- **sql-injection** indicates that this is considered a sink for the SQL injection query.

Continued example: Using type models
------------------------------------

Consider this variation on the previous example, the mysql2 EventMachine API is used.
The client is obtained via a call to **Mysql2::EM::Client.new**.

.. code-block:: ruby

  def submit(client, q)
    client = Mysql2::EM::Client.new
    client.query(q)
  end

So far we have only one model for **Mysql2::Client**, but in the real world we
may have many models for the various methods available. Because **Mysql2::EM::Client** is a subclass of **Mysql2::Client**, it inherits all of the same methods.
Instead of updating all our models to include both classes, we can add a type
model to indicate that **Mysql2::EM::Client** is a subclass of **Mysql2::Client**:

.. code-block:: yaml

  extensions:
    - addsTo:
        pack: codeql/ruby-all
        extensible: typeModel
      data:
        - ["Mysql2::Client", "Mysql2::EM::Client", ""]

Example: Adding flow through 'URI.decode_uri_component'
-------------------------------------------------------

> In this example, we'll show how to add flow through calls to `URI.decode_uri_component`:

Reviewer: ```suggestion
In this example, we'll show how to add flow through calls to 'URI.decode_uri_component':
```

Another one where I think it'll otherwise render in italics, and would be ideal to use the same format throughout the doc

.. code-block:: ruby

  y = URI.decode_uri_component(x); # add taint flow from 'x' to 'y'

We can model this using the following data extension:

.. code-block:: yaml

  extensions:
    - addsTo:
        pack: codeql/ruby-all
        extensible: summaryModel
      data:
        - [
            "URI",
            "Method[decode_uri_component]",
            "Argument[0]",
            "ReturnValue",
            "taint",
          ]

- Since we're adding flow through a method call, we add a tuple to the **summaryModel** extensible predicate.

> - The first column, **"URI!"**, begins the search for relevant calls at references to the **URI** class.

Reviewer: > The first column, **"URI!"**

Feel free to ignore but (from my layman POV) it looks like it says "URI" not "URI!"

- The **!** suffix indicates that we are looking for the class itself, rather than instances of the class.

> - The second column, **Member[decode_uri_component]**, is a path leading to the method calls we wish to model.

Reviewer: Should this be(?):

```suggestion
- The second column, **Method[decode_uri_component]**, is a path leading to the method calls we wish to model.
```

- Numeric operands to **Argument**, **Parameter**, and **Element** may be given as a lower bound. For example, **Argument[1..]** matches all arguments except 0.

Kinds
-----

Source kinds
~~~~~~~~~~~~

- **remote**: A generic source of remote flow. Most taint-tracking queries will use such a source. Currently this is the only supported source kind.

Sink kinds
~~~~~~~~~~

Unlike sources, sinks tend to be highly query-specific, rarely affecting more than one or two queries.
Not every query supports customizable sinks. If the following sinks are not suitable for your use case, you should add a new query.

- **code-injection**: A sink that can be used to inject code, such as in calls to **eval**.
- **command-injection**: A sink that can be used to inject shell commands, such as in calls to **child_process.spawn**.
- **path-injection**: A sink that can be used for path injection in a file system access, such as in calls to **fs.readFile**.
- **sql-injection**: A sink that can be used for SQL injection, such as in a MySQL **query** call.
- **url-redirection**: A sink that can be used to redirect the user to a malicious URL.

> - **log-injection**: A sink that can be used for log injection, such as in a **console.log** call.

Reviewer: Feel free to ignore as I'm not an SME, but just checking `console.log` is definitely applicable to Ruby and this wasn't a missed copy edit from the JS docs.
